### PR TITLE
cmd/k8s-operator,k8s-operator: add ProxyGroup CRD

### PR DESCRIPTION
**Pull request by tomhjp**
The ProxyGroup CRD specifies a set of N pods which will each be a tailnet device, and will have M different ingress or egress services mapped onto them. It is the mechanism for specifying how highly available proxies need to be. This commit only adds the definition, no controller loop, and so it is not currently functional.

Given that it doesn't have a controller loop implemented yet, this is probably incomplete, so it's very much just intended as a first draft.

Updates #13406

### CODE REVIEW

File: k8s-operator/apis/v1alpha1/types_proxygroup.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	// +optional
+	Replicas int `json:"replicas,omitempty"`
+
+	// ProxyClass is the name of the ProxyClass custom resource that
+	// contains configuration options that should be applied to the
+	// resources created for this ProxyGroup. If unset, the operator will
+	// create resources with the default configuration.
+	// +optional
+	ProxyClass string `json:"proxyClass,omitempty"`
+}
```
Comment (irbekrm) [inline k8s-operator/apis/v1alpha1/types_proxygroup.go:60]: Could we also add `Type` and `HostnamePrefix` fields in this PR?
See https://docs.google.com/document/d/1XQCaD2uTzeqEfOMgdxgGVsVOsRc7ojwtxheeOrSPwis/edit#bookmark=kix.3152s1kna2rs

File: k8s-operator/apis/v1alpha1/types_proxygroup.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+// +kubebuilder:printcolumn:name="Status",type="string",JSONPath=`.status.conditions[?(@.type == "ProxyGroupReady")].reason`,description="Status of the deployed ProxyGroup resources."
+
+type ProxyGroup struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec describes the desired ProxyGroup instances.
+	Spec ProxyGroupSpec `json:"spec"`
+
+	// ProxyGroupStatus describes the status of the ProxyGroup pods. This is
```
Comment (irbekrm) [inline k8s-operator/apis/v1alpha1/types_proxygroup.go:24]: Nit:
```suggestion
// ProxyGroupStatus describes the status of the ProxyGroup resources. This is
```

File: k8s-operator/apis/v1alpha1/types_proxygroup.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	// See  https://tailscale.com/kb/1236/kubernetes-operator/#setting-up-the-kubernetes-operator.
+	// Tags cannot be changed once a ProxyGroup device has been created.
+	// Tag values must be in form ^tag:[a-zA-Z][a-zA-Z0-9-]*$.
+	// +optional
+	Tags Tags `json:"tags,omitempty"`
+
+	// Replicas specifies how many replicas to create the StatefulSet with.
+	// Defaults to 2.
+	// +optional
+	Replicas int `json:"replicas,omitempty"`
```
Comment (irbekrm) [inline k8s-operator/apis/v1alpha1/types_proxygroup.go:52]: I think this should be a pointer to int.
We need to be able to distinguish between 'unset' and 'explicitly set to 0 (bc scale down for testing something or similar)'

File: k8s-operator/apis/v1alpha1/types_proxygroup.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+	// ProxyClass is the name of the ProxyClass custom resource that
+	// contains configuration options that should be applied to the
+	// resources created for this ProxyGroup. If unset, the operator will
```
Comment (irbekrm) [inline k8s-operator/apis/v1alpha1/types_proxygroup.go:56]:
```suggestion
// resources created for this ProxyGroup. If unset and no default ProxyClass is set, the operator will
```

File: k8s-operator/apis/v1alpha1/types_proxygroup.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	// create resources with the default configuration.
+	// +optional
+	ProxyClass string `json:"proxyClass,omitempty"`
+}
+
+type ProxyGroupStatus struct {
+	// List of status conditions to indicate the status of the ProxyGroup pods.
```
Comment (irbekrm) [inline k8s-operator/apis/v1alpha1/types_proxygroup.go:63]:
```suggestion
// List of status conditions to indicate the status of the ProxyGroup resources.
```

**Review by irbekrm — COMMENTED** [review]
Looks great! I've added a bunch of nits + I think we need to add at `Type` field and maybe also `HostnamePrefix` field to spec - but apart from that looks good to go in

File: k8s-operator/apis/v1alpha1/types_proxygroup.go (hunk 3 above)
Comment (tomhjp) [reply k8s-operator/apis/v1alpha1/types_proxygroup.go:52]: Good point, and ofc the "Defaults to 2" comment is only draft - I didn't have a strong opinion between 2 and 3.

File: k8s-operator/apis/v1alpha1/types_proxygroup.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// List of tailnet devices associated with the ProxyGroup statefulset.
+	// +listType=map
+	// +listMapKey=hostname
+	// +optional
+	Devices []TailnetDevice `json:"devices,omitempty"`
```
[the remark this answers is not in the document]
Comment (tomhjp) [reply k8s-operator/apis/v1alpha1/types_proxygroup.go:87]: Actually I realised we can't - the URL field is specific to recorders, so I had to split out a separate struct unfortunately.

File: k8s-operator/apis/v1alpha1/types_proxygroup.go [hunk 7]
```diff
@@ -49,27 +54,60 @@ type ProxyGroupSpec struct {
...
+	Hostname string `json:"hostname"`
+
+	// TailnetIPs is the set of tailnet IP addresses (both IPv4 and IPv6)
+	// assigned to the device.
+	// +optional
+	TailnetIPs []string `json:"tailnetIPs,omitempty"`
+}
+
+// +kubebuilder:validation:Type=string
+// +kubebuilder:validation:Pattern=`^[a-z0-9][a-z0-9-]{0,61}[a-z0-9]$`
```
Comment (irbekrm) [inline k8s-operator/apis/v1alpha1/types_proxygroup.go:103]: I think this should be an [enum](https://book.kubebuilder.io/reference/markers/crd-validation#:~:text=values%20specified%20here.-,//%20%2Bkubebuilder%3Avalidation%3AEnum,this%20(scalar)%20field%20is%20restricted%20to%20the%20*exact*%20values%20specified%20here.,-//%20%2Bkubebuilder%3Avalidation%3AExclusiveMaximum), we will only allow one of a specific set of values
Should be  `// +kubebuilder:validation:Enum=egress` (I assume we don't want to default to egress)

Comment (tomhjp) [reply k8s-operator/apis/v1alpha1/types_proxygroup.go]: Oops not sure how I managed that, could have sworn I had written it as an enum already. Thanks for catching.

**Review by irbekrm — APPROVED** [review]
Thanks @tomhjp , this looks great!

**Pull request merged**